### PR TITLE
Fixed: hashtable has a risk of division-by-zero

### PR DESCRIPTION
**Pull request by WereWind1**
During the static analysis we received a warning about potential division by zero on the hashtable container class. Further investigation pointed on the OnDeserialization method as likely the only potential risk factor among these warnings.

We have managed to reproduce the crash during the tests with deserializing a corrupted hashtable with BinaryFormatter class. The minimal code and both the valid and the invalid deserialized files for the crash reproduction are below.
 
```
using System.Runtime.Serialization.Formatters.Binary;
using System.Collections;

public class Program
{
        public static Hashtable load(string path)
        {
            FileStream inStr = new FileStream(path, FileMode.Open, FileAccess.Read,FileShare.Read);
            BinaryFormatter bf = new BinaryFormatter();
            Hashtable ht = bf.Deserialize(inStr) as Hashtable;
            inStr.Close();
            return ht;
        }

	public static void Main(string[] args)
	{
            Hashtable htgood = load("htemptygood.dat");
            htgood.Remove("1");
            Hashtable htbad = load("htemptybad.dat");
            htbad.Remove("1");
	}
}
```

[htempty.zip](https://github.com/user-attachments/files/15512569/htempty.zip)

Found by Linux Verification Center (linuxtesting.org) with SVACE.

### CODE REVIEW

**Comment by eiriktsarpalis** [comment]
Tagging @bartonjs @adamsitnik. How does this change factor into our plans to remove BF?

**Comment by stephentoub** [comment]
> with deserializing a corrupted hashtable with BinaryFormatter class

Corrupted how? There are many ways deserializing corrupted data could result in exceptions.

File: src/libraries/System.Private.CoreLib/src/System/Collections/Hashtable.cs [hunk 1]
```diff
@@ -1139,6 +1139,11 @@ public virtual void OnDeserialization(object? sender)
                 }
             }
 
+            if (hashsize < InitialSize)
```
[the remark this answers is not in the document]
Comment (WereWind1) [reply src/libraries/System.Private.CoreLib/src/System/Collections/Hashtable.cs:1142]: What would be the best place for the test? HashtableTests.cs or a new file specially for the issue?

[the remark this answers is not in the document]
Comment (eiriktsarpalis) [reply src/libraries/System.Private.CoreLib/src/System/Collections/Hashtable.cs:1142]: You could start by sharing a minimal reproduction in this thread, and we can have a conversation on whether a fix is warranted.

**Comment by WereWind1** [comment]
@stephentoub both the corrupted and uncorrupted files are in the archive attached to the first post. Basically what you need to do is serialize an empty hastable, change the size of the table from 3 (the default value for the empty table) to 0 and then deserialize the table back. 

Corrupting the non-empty hashtable this way doesn't lead to the exception

File: src/libraries/System.Private.CoreLib/src/System/Collections/Hashtable.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (WereWind1) [reply src/libraries/System.Private.CoreLib/src/System/Collections/Hashtable.cs:1142]:
```
using System.Collections;
using System.Diagnostics;
using System.Runtime.Serialization.Formatters.Binary;
namespace ConsoleApp1
{
    internal class Program
    {
        static void Main(string[] args)
        {
            Hashtable htable = new ();
            BinaryFormatter formatter = new ();
            MemoryStream stream = new ();
            formatter.Serialize(stream, htable);

            // make sure that the size is properly serialized
            Debug.Assert(stream.ToArray()[207] == 3);
            stream.Position = 207;

            //Change _loadsize into '0' or '1'
            stream.Write(new byte[] { 0 });
            //stream.Write(new byte[] { 1 });

            stream.Position = 0;
            var obj = (Hashtable) formatter.Deserialize(stream);

            // code throws exception on 'get' or 'remove'
            var tmp = obj[1];
            //obj.Remove(1);
        }
    }
}
```

**Comment by bartonjs** [comment]
Using BinaryFormatter is the same as using private reflection, so it puts objects into a state that they can't otherwise be in.  That's part of why BinaryFormatter has been marked as obsolete for years, and will be removed for .NET 9.

I'm a little confused about this OnDeserialization method.  It's public (for some reason), but the data it processes came from elsewhere (presumably a (de)serialization ctor called by BinaryFormatter. So I'm having trouble deciding if this is a "public API hardening" or a "private reflection hardening" change.

The change doesn't really hurt... but it's definitely hard to test (given BF's pending removal).

File: src/libraries/System.Private.CoreLib/src/System/Collections/Hashtable.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (jkotas) [reply src/libraries/System.Private.CoreLib/src/System/Collections/Hashtable.cs:1142]: It is not safe to use binary de-serialization against untrusted content. There are many crashes and security bugs that can be introduced by using binary de-serialization on corrupted or malicious payload. There is no point in trying to harden the implementation of binary deserialization against that.

**Pull request closed without merging**